packages/cli/src/scaling/scaling.service.ts

```diff
@@ -0,0 +1,220 @@
+} from './types';
+import type { IDataObject, IExecuteResponsePromiseData } from 'n8n-workflow';
+
+@Service()
+export class ScalingService {
+	private queue: JobQueue;
+
+	private readonly instanceType = config.getEnv('generic.instanceType');
+
+	constructor(
+		private readonly logger: Logger,
+		private readonly activeExecutions: ActiveExecutions,
+		private readonly processor: Processor,
+	) {}
+
+	/**
+	 * Lifecycle
+	 */
+
+	async setupQueue() {
+		const { default: BullQueue } = await import('bull');
+		const { RedisClientService } = await import('@/services/redis/redis-client.service');
+		const service = Container.get(RedisClientService);
+
+		const bullPrefix = config.getEnv('queue.bull.prefix');
+		const prefix = service.toValidPrefix(bullPrefix);
+
+		this.queue = new BullQueue(QUEUE_NAME, {
+			prefix,
+			settings: config.get('queue.bull.settings'),
+			createClient: (type) => service.createClient({ type: `${type}(bull)` }),
+		});
+
+		this.registerListeners();
+
+		this.logger.debug('[ScalingService] Queue setup completed');
+	}
+
+	setupWorker(concurrency: number) {
+		this.assertWorker();
+
+		void this.queue.process(
+			JOB_TYPE_NAME,
+			concurrency,
+			async (job: Job) => await this.processor.processJob(job),
+		);
+
+		this.logger.debug('[ScalingService] Worker setup completed');
+	}
+
+	@OnShutdown(HIGHEST_SHUTDOWN_PRIORITY)
+	async pauseQueue() {
+		await this.queue.pause(true, true);
+
+		this.logger.debug('[ScalingService] Queue paused');
+	}
+
+	async pingQueue() {
+		await this.queue.client.ping();
+	}
+
+	/**
+	 * Jobs
+	 */
+
+	async addJob(jobData: JobData, jobOptions: JobOptions) {
+		const { executionId } = jobData;
+
+		const job = await this.queue.add(JOB_TYPE_NAME, jobData, jobOptions);
+
+		this.logger.info(`[ScalingService] Added job ${job.id} (execution ${executionId})`);
+
+		return job;
+	}
+
+	async getJob(jobId: JobId) {
+		return await this.queue.getJob(jobId);
+	}
+
+	async findJobsByState(statuses: JobStatus[]) {
+		return await this.queue.getJobs(statuses);
+	}
+
+	async stopJob(job: Job) {
+		const props = { jobId: job.id, executionId: job.data.executionId };
+
+		try {
+			if (await job.isActive()) {
+				await job.progress(ABORT_JOB_SIGNAL);
+				this.logger.debug('[ScalingService] Stopped active job', props);
+				return true;
+			}
+
+			await job.remove();
+			this.logger.debug('[ScalingService] Stopped inactive job', props);
+			return true;
+		} catch (error: unknown) {
+			await job.progress(ABORT_JOB_SIGNAL);
+			this.logger.error('[ScalingService] Failed to stop job', { ...props, error });
+			return false;
+		}
+	}
+
+	/**
+	 * Listeners
+	 */
+
+	private registerListeners() {
+		this.queue.on('global:progress', (_job: Job, report: JobProgressReport) => {
+			if (report.kind === 'webhook-response') {
+				const { executionId, response } = report;
+				this.activeExecutions.resolveResponsePromise(
+					executionId,
+					this.decodeWebhookResponse(response),
+				);
+			}
+		});
```

Comment:
```suggestion
		this.queue.on(
			'global:progress',
			(_job: JobId, report: typeof ABORT_JOB_SIGNAL | JobProgressReport) => {
				if (typeof report !== 'number' && report.kind === 'webhook-response') {
					const { executionId, response } = report;
					this.activeExecutions.resolveResponsePromise(
						executionId,
						this.decodeWebhookResponse(response),
					);
				}
			},
		);
```

It would be cool to have `job.progress` be type-safe at some point.

Reply:
Stale as I migrated abort signal to being a message rather than a constant.